### ISSUE TITLE
React 16 Migration

### ISSUE THREAD

**Issue by shlomitc** [wix/wix-style-react#2313]
# React 16

## Strategy
Stay backward compatible with React 15 first, in one of the next major version we will drop support

## Work on React 16 branch
- this branch https://github.com/wix/wix-style-react/tree/migrate-to-react-16
- or this PR https://github.com/wix/wix-style-react/pull/2390/files

## Known issues
| # | scope | description | PR | status
| --- | --- | --- | --- | ---
| 2 | `<BulkSelectionConsumer/>` | component works but tests throws error  | | 
| 3 | `<DatePicker/>` | broken due to wrong `ref`s usage | [possible fix](https://github.com/wix/wix-style-react/pull/2378/commits/2dc33f494370017706f94fb829d7c9314dd3d485) | 

## Done
| # | scope | description | PR | status
| --- | --- | --- | --- | ---
| 1 | `<DataTable/>` | minor issue | https://github.com/wix/wix-style-react/pull/2382 | fixed
| 2 | `<Tooltip/>` | Fix potential null dereferencing | https://github.com/wix/wix-style-react/pull/2391 | fixed
| 2 | `<Tabs/>` | Modify Tooltip test to work with class component  | https://github.com/wix/wix-style-react/pull/2398 | fixed
| 3 | `Backoffice <Tooltip/>` | component works but tests are wrong  | removed in V6 | 
| 4 | `<RichTextArea/>` | broken - requires `slate` upgrade and code fixes | https://github.com/wix/wix-style-react/pull/2239 | will be deprecated, so no action needed

**Comment by shlomitc** [comment]
`BulkSelectionConsumer` - Tests fail due to React printing to console and our test setup fails tests that write to console. React warns that we throw inside render. Possible solutions:
- Delete these tests
- Change the component - Instead of throwing, render a descriptive message for the developer

**Comment by shlomitc** [comment]
`Backoffice Tooltip` https://github.com/wix/wix-style-react/blob/master/src/Backoffice/Tooltip/Tooltip.spec.js#L269
`show with delay and immediately hide` test (and possibly others too) is flaky due to wrong timeout implementation in https://github.com/InterAl/wait-for-cond/blob/master/index.js#L107
Should be fixed in `wait-for-cond`.

**Comment by shlomitc** [comment]
`DatePicker`
Looks like we’re using the `DayPickerInput` in a weird way: As part of its API, it accepts a Component to use as the input and props to pass to it. We currently pass it a function component (`_renderInput` method on `DatePicker`) that ignores the props that `DayPickerInput` creates for it (via merging `inputProps` with `ref` and its implementation of event handlers) - thus we actually degenerate most of what DayPickerInput does, and instead we render our own `Calendar` and handle the logic of show/hide there.

Bottom line - if we’re degenerating the `DayPickerInput`, maybe we can get rid of it entirely - just use our Input directly, and the “ref” warning will go away…

**Comment by shlomitc** [comment]
Some research on how to support 2 react versions:

https://blueprintjs.com/
Here's a link to a commit that reinstated support for 15+16:
https://github.com/palantir/blueprint/commit/9a93f0939868e4fb801cfbc41e7ae850295cda39
The gist:
- they use 2 CI configurations
- they have a mono-repo, so they have an extra `package.json` with the alternative versions of react and supporting libraries
- they control the tested React version via an ENV var

**Comment by mosheKerbel** [comment]
Components with failing tests after upgrade:
[] ColorInput
[] BulkSelection
[] Text
[] DatePicker
[] SidebarSectionItem
[] NestableList
[] Modal
[] Slider
[] ListItemAction
[] Range
[] MultiSelect
[] RichTextArea
[] DataTable
[] Input
[] PopoverMenu
[] SortableList
[] Some components are failing the Enzyme testkits smoketest

**Comment by shlomitc** [comment]
done
